plugin/ochttp/stats_test.go

```diff
@@ -19,9 +23,112 @@ func TestVarsInitialized(t *testing.T) {
+		if m == nil {
+			t.Errorf("measure not initialized: %d", i)
+		}
+	}
 	for i, v := range DefaultViews {
 		if v == nil {
 			t.Errorf("view not initialized: %d", i)
 		}
 	}
 }
+
+func httpHandler(statusCode, respSize int) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(statusCode)
+		body := make([]byte, respSize)
+		w.Write(body)
+	})
+}
+
+func updateMean(mean float64, sample, count int) float64 {
+	if count == 1 {
+		return float64(sample)
+	}
+	return mean + (float64(sample)-mean)/float64(count)
+}
+
+func TestHandlerStatsCollection(t *testing.T) {
+	for _, v := range DefaultViews {
+		v.Subscribe()
+	}
+
+	views := []string{
+		"opencensus.io/http/server/request_count",
+		"opencensus.io/http/server/latency",
+		"opencensus.io/http/server/request_bytes",
+		"opencensus.io/http/server/response_bytes",
+	}
+
+	// TODO: test latency measurements?
+	tests := []struct {
+		name, method, target                 string
+		count, statusCode, reqSize, respSize int
+	}{
+		{"get 200", "GET", "http://opencensus.io/request/one", 10, 200, 512, 512},
+		{"post 503", "POST", "http://opencensus.io/request/two", 5, 503, 1024, 16384},
+		{"no body 302", "GET", "http://opencensus.io/request/three", 2, 302, 0, 0},
+	}
+	totalCount, meanReqSize, meanRespSize := 0, 0.0, 0.0
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			body := bytes.NewBuffer(make([]byte, test.reqSize))
+			r := httptest.NewRequest(test.method, test.target, body)
+			w := httptest.NewRecorder()
+			h := &Handler{
+				NoTrace: true,
+				Handler: httpHandler(test.statusCode, test.respSize),
+			}
+
+			for i := 0; i < test.count; i++ {
+				h.ServeHTTP(w, r)
+				totalCount++
+				// Distributions do not track sum directly, we must
+				// mimic their behaviour to avoid rounding failures.
+				meanReqSize = updateMean(meanReqSize, test.reqSize, totalCount)
+				meanRespSize = updateMean(meanRespSize, test.respSize, totalCount)
+			}
+		})
+	}
+
+	for _, viewName := range views {
+		v := view.Find(viewName)
+		if v == nil {
+			t.Errorf("view not found %q", viewName)
+			continue
+		}
+		rows, err := v.RetrieveData()
+		if err != nil {
+			t.Error(err)
+			continue
+		}
+		if got, want := len(rows), 1; got != want {
+			t.Errorf("len(%q) = %d; want %d", viewName, got, want)
+			continue
+		}
+		data := rows[0].Data
+		var count int
+		var sum float64
+		switch data := data.(type) {
+		case *view.CountData:
+			count = int(*(*int64)(data))
+		case *view.DistributionData:
+			count = int(data.Count)
+			sum = data.Sum()
+		default:
+			t.Errorf("don't know how to handle data type: %v", data)
+		}
+		if got, want := count, totalCount; got != want {
+			t.Fatalf("%s = %d; want %d", viewName, got, want)
+		}
+		if got, want := sum, meanReqSize*float64(totalCount); viewName == "opencensus.io/http/server/request_bytes" && got != want {
```

Comment:
I think we can make this code easier to follow, with:
```go
switch viewName {
case "opencensus.io/http/server/request_bytes":
     if got, want := sum, meanReqSize*float64(totalCount); got != want {...}

case "opencensus.io/http/server/response_bytes":
     if got, want := sum, meanRespSize*float64(totalCount); got != want {...}

default:
      t.Errorf("Stray view %q", viewName)
}
```

Reply:
We can't have the default case be an error, since we look at more than these two views. But otherwise I agree this is neater, thanks :-)